src/test/java/uk/gov/ons/census/casesvc/service/RefusalProcessorTest.java

```diff
@@ -0,0 +1,92 @@
+import static uk.gov.ons.census.casesvc.testutil.DataUtils.getTestResponseManagementEvent;
+
+import java.util.Optional;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+import uk.gov.ons.census.casesvc.logging.EventLogger;
+import uk.gov.ons.census.casesvc.model.dto.RefusalDTO;
+import uk.gov.ons.census.casesvc.model.dto.ResponseManagementEvent;
+import uk.gov.ons.census.casesvc.model.entity.Case;
+import uk.gov.ons.census.casesvc.model.entity.EventType;
+import uk.gov.ons.census.casesvc.model.entity.UacQidLink;
+import uk.gov.ons.census.casesvc.model.repository.CaseRepository;
+import uk.gov.ons.census.casesvc.model.repository.UacQidLinkRepository;
+
+@RunWith(MockitoJUnitRunner.class)
+public class RefusalProcessorTest {
+
+  private static final String REFUSAL_RECEIVED = "Refusal Received";
+
+  @Mock private UacQidLinkRepository uacQidLinkRepository;
+  @Mock private CaseRepository caseRepository;
+
+  @Mock private CaseProcessor caseProcessor;
+
+  @Mock private EventLogger eventLogger;
+
+  @InjectMocks RefusalProcessor underTest;
+
+  @Test
+  public void shouldProcessARefusalReceivedMessageSuccessfully() {
+    // GIVEN
+    ResponseManagementEvent managementEvent = getTestResponseManagementEvent();
+    Case testCase = getRandomCase();
+    testCase.setRefusalReceived(false);
+    UacQidLink expectedUacQidLink = testCase.getUacQidLinks().get(0);
+    Case expectedCase = expectedUacQidLink.getCaze();
+    RefusalDTO expectedRefusal = managementEvent.getPayload().getRefusal();
+
+    when(uacQidLinkRepository.findByQid(anyString())).thenReturn(Optional.of(expectedUacQidLink));
+
+    // WHEN
+    underTest.processRefusal(managementEvent);
+
+    // THEN
+    ArgumentCaptor<Case> caseArgumentCaptor = ArgumentCaptor.forClass(Case.class);
+    verify(caseRepository).saveAndFlush(caseArgumentCaptor.capture());
+    Case actualCase = caseArgumentCaptor.getValue();
+
+    assertThat(actualCase.isRefusalReceived()).isTrue();
+
+    verify(caseProcessor, times(1)).emitCaseUpdatedEvent(expectedCase);
+    verify(eventLogger, times(1))
+        .logRefusalEvent(
+            expectedUacQidLink,
+            REFUSAL_RECEIVED,
+            EventType.CASE_UPDATED,
+            expectedRefusal,
+            managementEvent.getEvent(),
+            expectedRefusal.getResponseDateTime());
+  }
+
+  @Test
+  public void shouldThrowRuntimeExceptionWhenCaseNotFound() {
+    // GIVEN
+    ResponseManagementEvent managementEvent = getTestResponseManagementEvent();
+    String expectedQuestionnaireId = managementEvent.getPayload().getRefusal().getQuestionnaireId();
+    String expectedErrorMessage =
+        String.format("Questionnaire Id '%s' not found!", expectedQuestionnaireId);
+
+    when(uacQidLinkRepository.findByQid(anyString())).thenReturn(Optional.empty());
+
+    try {
```

Comment:
This doesn't work if no Exception is thrown.  There's a few options, the magic @Rule:

```
  @Rule
  public ExpectedException expectedEx = ExpectedException.none();

  @Test
  public void shouldThrowRuntimeExceptionWhenCaseNotFound() {

    // GIVEN
    ResponseManagementEvent managementEvent = getTestResponseManagementEvent();
    String expectedQuestionnaireId = managementEvent.getPayload().getRefusal().getQuestionnaireId();
    String expectedErrorMessage =
        String.format("Questionnaire Id '%s' not found!", expectedQuestionnaireId);
    expectedEx.expectMessage(expectedErrorMessage);
    expectedEx.expect(RuntimeException.class);

    when(uacQidLinkRepository.findByQid(anyString())).thenReturn(Optional.empty());

    //when
    underTest.processRefusal(managementEvent);
  }

```

The link below has some good answers,  I'm not 100% onboard with the @Rule above as it feels a bit magical.   An answer from that link seems easier to follow to me?

```
@Test (expected = RuntimeException.class)
public void myTestMethod()
{
   try
   {
      //Run exception throwing operation here
   }
   catch(RuntimeException re)
   {
      String message = "Employee ID is null";
      assertEquals(message, re.getMessage());
      throw re;
    }
    fail("Employee Id Null exception did not throw!");
  }

```

https://stackoverflow.com/questions/2469911/how-do-i-assert-my-exception-message-with-junit-test-annotation

Reply:
changed it to be consistent with refusal